keras_core/layers/layer.py

```diff
@@ -740,9 +740,7 @@ def maybe_convert(x):
         # 7. Call the layer.
         try:
             with backend.name_scope(self.name, caller=self):
-                if self.autocast and self.compute_dtype != self.variable_dtype:
-                    # For mixed precision, we automatically cast layer variables
-                    # (float ones only) to the compute dtype upon access.
+                if self.autocast and backend.is_float_dtype(self.compute_dtype):
```

Comment:
We should seek to avoid creating an `AutocastScope` if unnecessary because of the eager performance overhead (which matters specifically for torch).

To handle nesting I guess we can further complicate the logic, something like

```python
if self.autocast and backend.is_float_dtype(self.compute_dtype):
    autocast_scope = get_autocast_scope()
    if (autocast_scope and autocast_scope.dtype != self.compute_dtype) or (not autocast_scope and self.compute_dtype != self.variable_dtype):
        with backend.AutocastScope(self.compute_dtype):
```

WDYT? Too much? Is the performance cost of many AutocastScopes low enough we can go with your proposal?

Reply:
Sounds good. If we want to be fully correct, I think we actually need to add the concept of an `empty scope`. E.g. if layer with `autocast = False` is inside a model, we actually need to "clear the scope" which will have been set.